utils/active-hotspot-oracle/src/main.rs

```diff
@@ -0,0 +1,489 @@
+      .context("No beacon_pub_key column array found")?
+      .as_any()
+      .downcast_ref::<StringArray>()
+      .context("Result is not a string array")?;
+    let witness_pub_keys = record_batch
+      .column_by_name("witness_pub_key")
+      .context("No witness_pub_key column array found")?
+      .as_any()
+      .downcast_ref::<StringArray>()
+      .context("Result is not a string array")?;
+
+    count += beacon_pub_keys.len();
+    for pk in beacon_pub_keys {
+      let pub_key = pk.unwrap_or("").clone().to_string();
+      active_pub_keys.insert(pub_key);
+    }
+
+    for pk in witness_pub_keys {
+      let pub_key = pk.unwrap_or("").clone().to_string();
+      active_pub_keys.insert(pub_key);
+    }
+  }
+
+  println!("Records processed: {:?}", count);
+  println!("Active pub keys: {:?}", active_pub_keys.len());
+
+  // download last checkpoint file from s3
+  println!("Loading last save files from s3");
+  let last_active_pub_keys: HashSet<String> =
+    load_checkpoint(s3_config.clone(), "active_save.json")
+      .await
+      .unwrap_or(HashSet::new());
+  let mut last_inactive_pub_keys: HashSet<String> =
+    load_checkpoint(s3_config.clone(), "inactive_save.json")
+      .await
+      .unwrap_or(HashSet::new());
+
+  // parse and find the diff
+  println!("Finding new diffs");
+  let mark_active_diff: Vec<String> = active_pub_keys
+    .difference(&last_active_pub_keys)
+    .cloned()
+    .collect();
+  let mark_inactive_diff: Vec<String> = last_active_pub_keys
+    .difference(&active_pub_keys)
+    .cloned()
+    .collect();
+
+  // construct the transactions for the diffs
+  println!("Constructing instructions");
+  let (mark_active_iot_ixs, mark_active_mobile_ixs) =
+    construct_set_active_ixs(&helium_entity_program, &mark_active_diff, true)?;
+  let (mark_inactive_iot_ixs, mark_inactive_mobile_ixs) =
+    construct_set_active_ixs(&helium_entity_program, &mark_inactive_diff, false)?;
+
+  // send transactions
+  println!("Sending mark active iot transactions");
+  construct_and_send_txs(
+    &helium_entity_program,
+    &tpu_client,
+    mark_active_iot_ixs,
+    &kp,
+  )?;
+  println!("Sending mark active mobile transactions");
+  construct_and_send_txs(
+    &helium_entity_program,
+    &tpu_client,
+    mark_active_mobile_ixs,
+    &kp,
+  )?;
+  println!("Sending mark inactive iot transactions");
+  construct_and_send_txs(
+    &helium_entity_program,
+    &tpu_client,
+    mark_inactive_iot_ixs,
+    &kp,
+  )?;
+  println!("Sending mark inactive mobile transactions");
+  construct_and_send_txs(
+    &helium_entity_program,
+    &tpu_client,
+    mark_inactive_mobile_ixs,
+    &kp,
+  )?;
+
+  // write to new checkpoint file and upload to s3
+  println!("Uploading new save files to s3");
+  last_inactive_pub_keys.extend(mark_inactive_diff.iter().cloned());
+  let active_pub_keys = HashSet::new();
+  save_checkpoint(active_pub_keys, s3_config.clone(), "active_save.json").await?;
+  save_checkpoint(
```

Comment:
Is there a race condition here if saving a checkpoint fails? Now we have on-chain not matching the checkpoint of actives/inactives.

Reply:
If saving the checkpoint fails then the next time the job runs it will run from the previous checkpoint. In this application we need to guarantee a tx will run at least once, if a tx gets run twice it's better than skipping some. 

Should be using the fetched infos to check if a tx even needs to be run though.

Reply:
Couldn't the following scenario break us:

  Last checkpoint says hotspot A is active
  Currently A is not active
  Update A to not active
  Save checkpoint fails
  
  Next run:
  Last checkpoint says A is active
  Currently A is active (it was turned back on again)
  A matches last status, so no change. (BUG)

Reply:
It doesn't "break", that scenario doesn't cause any unintended consequences and I wrote it with that possibility in mind. But yes, as I said in my last comment I will change it to use the fetched info account to determine if the tx needs to be run or not to avoid wasting a tx.